// Tinder-App/src/environments/environment.ts
// This file can be replaced during build by using the `fileReplacements` array.
// `ng build` replaces `environment.ts` with `environment.prod.ts`.
// The list of file replacements can be found in `angular.json`.

export const environment = {
  production: false,

  firebase: {
    apiKey: "AIzaSyCx8BsIuD8eDQ0TxLjqCFA5CTGY3kd3JAc",
    authDomain: "tinder-app-8acfc.firebaseapp.com",
    databaseURL: "https://tinder-app-8acfc-default-rtdb.firebaseio.com/",
    projectId: "tinder-app-8acfc",
    storageBucket: "tinder-app-8acfc.firebasestorage.app",
    messagingSenderId: "41918120775",
    appId: "1:41918120775:web:e28a1be84027b022f04262",
    measurementId: "G-JBMVSKCL3K"
  },
  enablePush: true,
};

/*
 * For easier debugging in development mode, you can import the following file
 * to ignore zone related error stack frames such as `zone.run`, `zoneDelegate.invokeTask`.
 *
 * This import should be commented out in production mode because it will have a negative impact
 * on performance if an error is thrown.
 */
// import 'zone.js/plugins/zone-error';  // Included with Angular CLI.
